packages/lodestar/src/network/encoders/request.ts

```diff
@@ -1,42 +1,59 @@
-import {Method, Methods, ReqRespEncoding} from "../../constants";
+import pipe from "it-pipe";
+import BufferList from "bl";
+import {decode, encode} from "varint";
 import {ILogger} from "@chainsafe/lodestar-utils";
 import {RequestBody} from "@chainsafe/lodestar-types";
-import {decode, encode} from "varint";
-import BufferList from "bl";
 import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {Method, Methods, ReqRespEncoding} from "../../constants";
+import {toBuffer} from "../../util/buffer";
 import {getCompressor, getDecompressor, maxEncodedLen} from "./utils";
 import {IValidatedRequestBody} from "./interface";
 
-export function eth2RequestEncode(
+export async function streamRequestBodyTo(
   config: IBeaconConfig,
-  logger: ILogger,
   method: Method,
-  encoding: ReqRespEncoding
-): (source: AsyncIterable<RequestBody | null>) => AsyncGenerator<Buffer> {
-  return async function* (source) {
-    const type = Methods[method].requestSSZType(config);
-    let compressor = getCompressor(encoding);
+  encoding: ReqRespEncoding,
+  requestBody: RequestBody,
+  requestBodySink: Stream
+): Promise<void> {
+  if (requestBody == null) return;
 
-    for await (const request of source) {
-      if (!type || request === null) continue;
+  const serialized = serializeRequestBody(config, method, requestBody);
+  if (serialized == null) return;
 
-      let serialized: Uint8Array;
-      try {
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        serialized = type.serialize(request as any);
-      } catch (e) {
-        logger.warn("Malformed input. Failed to serialize request", {method}, e);
-        continue;
-      }
-
-      yield Buffer.from(encode(serialized.length));
+  const bodyEncodedStream = getRequestBodyEncodedStream(encoding, serialized);
+  await pipe(bodyEncodedStream, requestBodySink);
+}
 
-      yield* compressor(Buffer.from(serialized.buffer, serialized.byteOffset, serialized.length));
+/**
+ * Serializes `requestBody` to bytes serialization of `method` SSZ type
+ * SSZ type for `method` may not exist, then returns null
+ */
+export function serializeRequestBody(
+  config: IBeaconConfig,
+  method: Method,
+  requestBody: RequestBody
+): Uint8Array | null {
+  const type = Methods[method].requestSSZType(config);
+  if (type == null || requestBody == null) {
+    return null;
+  } else {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return type.serialize(requestBody as any);
```

Comment:
I feel like catching error here and logging it makes more sense. Currently it would end up as unhandled rejection and we won't close streams we opened.

Reply:
Added a commit to close stream on error